src/shared/tpm2-util.c

```diff
@@ -7905,45 +7928,58 @@ Tpm2Support tpm2_support_full(Tpm2Support mask) {
                 support |= TPM2_SUPPORT_SYSTEM;
 
         if (FLAGS_SET(mask, TPM2_SUPPORT_LIBRARIES)) {
-                r = dlopen_tpm2();
+                r = dlopen_tpm2_esys();
                 if (r >= 0)
+                        support |= TPM2_SUPPORT_LIBTSS2_ESYS;
+
+                r = dlopen_tpm2_rc();
+                if (r >= 0)
+                        support |= TPM2_SUPPORT_LIBTSS2_RC;
+
+                r = dlopen_tpm2_mu();
+                if (r >= 0)
+                        support |= TPM2_SUPPORT_LIBTSS2_MU;
+
+                if (FLAGS_SET(support, TPM2_SUPPORT_LIBTSS2_ALL))
                         support |= TPM2_SUPPORT_LIBRARIES;
         }
 #endif
 
         return support;
 }
 
+static void print_field(const char *prefix, const char *s, bool supported) {
+        printf("%s%s%s%s%s\n", strempty(prefix), supported ? ansi_green() : ansi_red(), plus_minus(supported), s, ansi_normal());
+}
+
 int verb_has_tpm2_generic(bool quiet) {
         Tpm2Support s;
 
         s = tpm2_support();
 
         if (!quiet) {
                 if (s == TPM2_SUPPORT_FULL)
-                        puts("yes");
+                        printf("%syes%s\n", ansi_green(), ansi_normal());
                 else if (s == TPM2_SUPPORT_NONE)
-                        puts("no");
+                        printf("%sno%s\n", ansi_red(), ansi_normal());
                 else
-                        puts("partial");
-
-                printf("%sfirmware\n"
-                       "%sdriver\n"
-                       "%ssystem\n"
-                       "%ssubsystem\n"
-                       "%slibraries\n",
-                       plus_minus(s & TPM2_SUPPORT_FIRMWARE),
-                       plus_minus(s & TPM2_SUPPORT_DRIVER),
-                       plus_minus(s & TPM2_SUPPORT_SYSTEM),
-                       plus_minus(s & TPM2_SUPPORT_SUBSYSTEM),
-                       plus_minus(s & TPM2_SUPPORT_LIBRARIES));
+                        printf("%spartial%s\n", ansi_yellow(), ansi_normal());
+
+                print_field(NULL, "firmware", FLAGS_SET(s, TPM2_SUPPORT_FIRMWARE));
+                print_field(NULL, "driver", FLAGS_SET(s, TPM2_SUPPORT_DRIVER));
+                print_field(NULL, "system", FLAGS_SET(s, TPM2_SUPPORT_SYSTEM));
+                print_field(NULL, "subsystem", FLAGS_SET(s, TPM2_SUPPORT_SUBSYSTEM));
+                print_field(NULL, "libraries", FLAGS_SET(s, TPM2_SUPPORT_LIBRARIES));
+                print_field("  ", "libtss2-esys.so.0", FLAGS_SET(s, TPM2_SUPPORT_LIBTSS2_ESYS));
+                print_field("  ", "libtss2-rc.so.0", FLAGS_SET(s, TPM2_SUPPORT_LIBTSS2_RC));
+                print_field("  ", "libtss2-mu.so.0", FLAGS_SET(s, TPM2_SUPPORT_LIBTSS2_MU));
```

Comment:
honestly, I am not convinced we should show these. to me these are an implementation detail about the backend stack we use. We have currently no reason to swap out the backend stack, but I don't think we should "leak" this to users. 

Also, these three libraries to my knowledge are never packaged separately, and the esys one has a dep on th eother ones anyway, hence the fine grained info seems kinda redundant: it's not that you will have the esys one installed but the mu one not, with any of today's package managers.

I think it would be OK to maybe mention this in the man pages, but then again afaik we don't document the names of the other libs we dlopen() like that, so I probably wouldn't bother, but I also don't mind if you add that.

Reply:
But, systemd-analyze is a kind of debugging tool. I think it is OK to show an 'implementation detail'.
Honestly, I'd like to show more, e.g. loaded drivers, how we detects if firmware supports TPM2 and so on.